File: frontend/src/lib/schemas/payment.ts
import { z } from "zod";

const currencySchema = z
  .string()
  .trim()
  .toUpperCase()
  .regex(/^[A-Z]{3}$/u, "Currency must be a 3-letter ISO code");

export const moneySchema = z.object({
  currency: currencySchema,
  value_minor: z
    .number({ required_error: "Amount is required" })
    .int("Amount must be an integer")
    .nonnegative("Amount must be positive"),
});

export const paymentResponseSchema = z.object({
  transaction_id: z.string(),
  status: z.string(),
  reason: z.string().nullable().optional(),
});

export const paymentRecordSchema = z.object({
  transaction_id: z.string(),
  merchant_id: z.string(),
  customer_id: z.string(),
  amount: moneySchema,
  payment_method: z.string(),
  status: z.string(),
  reason: z.string().nullable().optional(),
  created_at: z.string().min(1),
  updated_at: z.string().min(1),
});

export const paymentCollectionSchema = z.object({
  items: z.array(paymentRecordSchema),
  total: z.number().int().nonnegative(),
});

export const paymentRequestSchema = z.object({
  transaction_id: z.string().min(8).max(64),
  merchant_id: z.string().min(4).max(64),
  customer_id: z.string().min(4).max(64),
  amount: moneySchema.refine((value) => value.value_minor > 0, {
    message: "Amount must be greater than zero",
  }),
  payment_method: z.string().min(2).max(32),
  reference: z
    .string()
    .max(128)
    .optional()
    .transform((value) => (value ? value.trim() : undefined)),
});

export const paymentFormSchema = z.object({
  transactionId: z.string().min(8).max(64),
  merchantId: z.string().min(4).max(64),
  customerId: z.string().min(4).max(64),
  amountMinor: z
    .coerce.number({ invalid_type_error: "Amount must be a number" })
    .int("Amount must be an integer")
    .positive("Amount must be greater than zero"),
  currency: currencySchema.default("INR"),
  paymentMethod: z.string().min(2).max(32),
  reference: z
    .string()
    .max(128)
    .optional()
    .transform((value) => (value ? value.trim() : undefined)),
});

export type PaymentRequestPayload = z.infer<typeof paymentRequestSchema>;
export type PaymentResponse = z.infer<typeof paymentResponseSchema>;
export type PaymentRecord = z.infer<typeof paymentRecordSchema>;
export type PaymentCollection = z.infer<typeof paymentCollectionSchema>;
export type PaymentFormValues = z.infer<typeof paymentFormSchema>;

File: services/gateway_api/gateway_api/schemas/payment.py
from datetime import datetime
from typing import Annotated, Optional

from pydantic import BaseModel, Field, StringConstraints


CurrencyStr = Annotated[str, StringConstraints(min_length=3, max_length=3)]
TransactionIdStr = Annotated[str, StringConstraints(min_length=8, max_length=64)]
PartyIdStr = Annotated[str, StringConstraints(min_length=4, max_length=64)]
PaymentMethodStr = Annotated[str, StringConstraints(min_length=2, max_length=32)]


class Money(BaseModel):
    currency: CurrencyStr = Field(..., description="ISO 4217 currency code")
    value_minor: int = Field(..., gt=0)


class PaymentRequest(BaseModel):
    transaction_id: TransactionIdStr
    merchant_id: PartyIdStr
    customer_id: PartyIdStr
    amount: Money
    payment_method: PaymentMethodStr
    reference: Optional[str] = Field(default=None, max_length=128)


class PaymentResponse(BaseModel):
    transaction_id: str
    status: str
    reason: Optional[str] = None


class PaymentRecord(BaseModel):
    transaction_id: str
    merchant_id: str
    customer_id: str
    amount: Money
    payment_method: str
    status: str
    reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class PaymentCollection(BaseModel):
    items: list[PaymentRecord]
    total: int
